### PR TITLE
Computed Columns

### LINKED ISSUE #2054: Support materialized computed columns for primary keys
Background: https://airtable.com/tblD3oZPLJgGhCmch/viwINAgXA5on9pr2T/recz17DReeMym5Wyv

PRs:

### PR DESCRIPTION
**Pull request by lhirata**
Closes #2054.

Links to partitioning docs will be added in later (See #2764).

### CODE REVIEW

**Comment by justinj** [comment]
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/jsonb.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DxsS8hzqsBjtdlJ4y:-L8DxsS8hzqsBjtdlJ4z:bptpl4y) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/jsonb.md#L19)):*
> ```Markdown
>     ('{"name": "Buster Bunny", "age": "15", "id": "f98112", "school": "THS", "credits": 67, "clubs": "MUN"}'),
>     ('{"name": "Ernie Dog", "school" : "Brooklyn Tech", "id": "t63512", "sports": "Track and Field", "clubs": "Chess"}'
>   );
> ```

nit: I think this paren should go on the preceding line?

---

*[_includes/computed-columns/partitioning.md, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dy43tP4uQYBCz53uM:-L8Dy43tP4uQYBCz53uN:b-vdhdo0) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L31)):*
> ```Markdown
>     ('Scotty', 'mx'),
>     ('Hikaru Sulu', 'us'),
>     ('Pavel Chekov', 'au'
> ```

nit: same here

---

*[_includes/computed-columns/simple.md, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DyPbM53eRbM62Ardi:-L8DyPbOncvKkiIOAQR1:b-iipjm8) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/simple.md#L20)):*
> ```Markdown
>     (1, 'Lola', 'McDog'),
>     (2, 'Carl', 'Kimball'),
>     (3, 'Ernie', 'Floof'
> ```

nit: ditto

---

*[v2.0/computed-columns.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dz5cvxLp4JQUGuDre:-L8Dz5cvxLp4JQUGuDrf:b-z0z5en) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L13)):*
> ```Markdown
> ## Why use computed columns?
> 
> Computed columns are especially useful when used with partitioning or [`JSONB`](jsonb.html) columns.
> ```

could also do a link to partitioning maybe?

---

*[v2.0/computed-columns.md, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzIOztmUQxE-VAdNw:-L8DzIOztmUQxE-VAdNx:b-8p63nj) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L24)):*
> ```Markdown
> 
> - Cannot be added once a table is created.
> - Cannot reference other computer columns.
> ```

nit: computed, and maybe clarify what's meant by "reference" here, because that word sort of comes with the connotation of foreign keys I think? not sure what something better would be, though.

---

*[v2.0/computed-columns.md, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8E-g7Wtg_MO8I5dj5-:-L8E-g7Wtg_MO8I5dj50:bpbng0a) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L30)):*
> ```Markdown
> ## Creation
> 
> Computed columns can only be added at the time of [table creation](create-table.html). Use the following syntax:
> ```

we have a tracking issue for this: https://github.com/cockroachdb/cockroach/issues/22652 do we want to link people here to track demand? don't know if there's any precedent for that

---

*[v2.0/computed-columns.md, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzXyIScTCdQUbDxDZ:-L8DzXyIScTCdQUbDxD_:bjzonmw) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L33)):*
> ```Markdown
> 
> ~~~
> column_name <type> AS <expr> [ STORED | VIRTUAL]
> ```

We only support `STORED`, though the RFC mentioned virtual columns, this was aspirational. also we should wrap `<expr>` in parens: `(<expr>)`.

---

*[v2.0/computed-columns.md, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzerWP_KoX-7a8efw:-L8DzerWP_KoX-7a8efx:b1iutzp) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L41)):*
> ```Markdown
> `<type>` | The [data type](data-types.html) of the computed column.
> `<expr>` | The pure [expression](sql-expressions.html) used to compute column values. Any functions marked as `impure`, such as `now()` or `nextval()` cannot be used.
> `STORED` | _(Default)_ The computed column is stored alongside other columns. This is required if the column appears in the [Primary Key](primary-key.html).
> ```

it's actually not the default - it's required.

---

*[v2.0/computed-columns.md, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dzs1IKqde8FsnRVzs:-L8Dzs1IKqde8FsnRVzt:brbjgqh) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L42)):*
> ```Markdown
> `<expr>` | The pure [expression](sql-expressions.html) used to compute column values. Any functions marked as `impure`, such as `now()` or `nextval()` cannot be used.
> `STORED` | _(Default)_ The computed column is stored alongside other columns. This is required if the column appears in the [Primary Key](primary-key.html).
> `VIRTUAL` | The computed column is only computed when rows are read.
> ```

we don't support this (yet) :(

---

*[v2.0/inverted-indexes.md, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8E-4_r4ngdMqMKVngP:-L8E-4_r4ngdMqMKVngQ:bjcs58f) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/inverted-indexes.md#L115)):*
> ```Markdown
> {% include copy-clipboard.html %}
> ~~~ sql
> > SELECT * jsonb_pretty(user_profile) FROM users;
> ```

does this work? before or after this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2759)*

**Comment by lhirata** [comment]
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/jsonb.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DxsS8hzqsBjtdlJ4y:-L8EBz8mrc2KyPWXh-Yk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/jsonb.md#L19)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: I think this paren should go on the preceding line?
</blockquote></details>

Done.

---

*[_includes/computed-columns/partitioning.md, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dy43tP4uQYBCz53uM:-L8EC1cfO_HYHrIPbmUE:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L31)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: same here
</blockquote></details>

Done.

---

*[_includes/computed-columns/simple.md, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DyPbM53eRbM62Ardi:-L8EC4fozR_UfZuzc_w7:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/simple.md#L20)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: ditto
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dz5cvxLp4JQUGuDre:-L8EC5moZa2L2oVWSmGk:b-6o06r4) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L13)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

could also do a link to partitioning maybe?
</blockquote></details>

will add the link once amruta publishes partitioning (i have it tracked in a different issue)

---

*[v2.0/computed-columns.md, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzIOztmUQxE-VAdNw:-L8ED-eUnIwV68k7fERn:b-jxzc51) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L24)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: computed, and maybe clarify what's meant by "reference" here, because that word sort of comes with the connotation of foreign keys I think? not sure what something better would be, though.
</blockquote></details>

hmm, not sure what would be better either. something like "cannot be used to compute other computed columns"?

---

*[v2.0/computed-columns.md, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzXyIScTCdQUbDxDZ:-L8EDhrS-zGyiaIhExjL:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L33)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

We only support `STORED`, though the RFC mentioned virtual columns, this was aspirational. also we should wrap `<expr>` in parens: `(<expr>)`.
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8DzerWP_KoX-7a8efw:-L8EEWV3cfQkuLkPtKfN:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L41)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

it's actually not the default - it's required.
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dzs1IKqde8FsnRVzs:-L8EEWyZ6ZulYY6ydMIC:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L42)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

we don't support this (yet) :(
</blockquote></details>

Done.

---

*[v2.0/inverted-indexes.md, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8E-4_r4ngdMqMKVngP:-L8EBaQlt3nFSFtH5Zrb:b-j9uxhl) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/inverted-indexes.md#L115)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

does this work? before or after this change?
</blockquote></details>

you're right - it doesn't work now. i thought the `,` was a typo that i missed. reverted the change

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2759)*

**Comment by lhirata** [comment]
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[v2.0/computed-columns.md, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8E-g7Wtg_MO8I5dj5-:-L8EL9eK3xN0QC54xVJb:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L30)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

we have a tracking issue for this: https://github.com/cockroachdb/cockroach/issues/22652 do we want to link people here to track demand? don't know if there's any precedent for that
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2759)*

**Comment by jseldess** [comment]
*[v2.0/computed-columns.md, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8SU417J0NbMoRs2ATZ:-L8SU417J0NbMoRs2AT_:b-tut2ds) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L15)):*
> ```Markdown
> Computed columns are especially useful when used with partitioning or [`JSONB`](jsonb.html) columns.
> 
> **Partitioning** requires that partitions are defined using columns that are a prefix of the [primary key](primary-key.html). In the case of geo-partitioning, some applications will want to collapse the number of possible values in this column, to make certain classes of queries more performant. For example, if a users table has a country and state column, then a CockroachDB operator could make a stored computed column locality with a reduced domain for use in partitioning. For more information, see the [partitioning example](#create-a-table-with-geo-partitions-and-a-computed-column) below.
> ```

Also, maybe make this a bulleted list:

```
- **Partitioning** ...

- **`JSONB`** ...
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2759)*

**Comment by jseldess** [comment]
:lgtm: with a few comments. Thanks, @lhirata!

---

Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/partitioning.md, line 2 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8SVX-kf0UQxHWpI8et:-L8SVX-kf0UQxHWpI8eu:b99h2dq) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L2)):*
I think we need a callout that this is an enterprise feature and thus requies an enterprise licesne. Something like:

> Partitioning is an enterprise feature. To request and enable a trial or full enterprise license, see [Enterprise Licensing](enterprise-licensing.html). 

---

*[_includes/computed-columns/partitioning.md, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8Dy43tP4uQYBCz53uM:-L8SWG_wQZAJyHWkwS10:bptuuoa) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L31)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Parens are still on a separate line, @lhirata.

---

*[v2.0/computed-columns.md, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8SSpOT3g4f0INeATee:-L8SSpOT3g4f0INeATef:bo8tz96) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L15)):*
> ```Markdown
> Computed columns are especially useful when used with partitioning or [`JSONB`](jsonb.html) columns.
> 
> **Partitioning** requires that partitions are defined using columns that are a prefix of the [primary key](primary-key.html). In the case of geo-partitioning, some applications will want to collapse the number of possible values in this column, to make certain classes of queries more performant. For example, if a users table has a country and state column, then a CockroachDB operator could make a stored computed column locality with a reduced domain for use in partitioning. For more information, see the [partitioning example](#create-a-table-with-geo-partitions-and-a-computed-column) below.
> ```

I'd prefer "you" over "a CockroachDB operator".

Also, in the last sentence, I don't understand `a stored computed column locality`. Can you rephrase that so that it's easier to unpack?

---

*[v2.0/jsonb.md, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8SXETsG8eNRZdtNziP:-L8SXETsG8eNRZdtNziQ:bhlx6la) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/jsonb.md#L188)):*
> ```Markdown
> ### Create a table with a `JSONB` column and a computed column
> 
> {% include computed-columns/jsonb.md %}
> ```

You might want to change the table name in this included example from `students` to something else. `students` is already used on this page, and so copy/pasting the `create table` in this example gives you an error:

```
root@:26257/test> CREATE TABLE students (                                                                                                                                             id STRING PRIMARY KEY AS (profile->>'id') STORED,                                                                                                                                 profile JSONB                                                                                                                                                                 );
pq: relation "students" already exists
```

---

*[v2.0/jsonb.md, line 9 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-L8SXrp7iQsyvnEx6fho:-L8SXrp7iQsyvnEx6fhp:b-m3mp80) ([raw file](https://github.com/cockroachdb/docs/blob/830a7e8ca0ca5c1af05f5d088956e06833a3be8a/v2.0/jsonb.md#L9)):*
> ```Markdown
> <span class="version-tag">New in v2.0:</span> The `JSONB` [data type](data-types.html) stores JSON (JavaScript Object Notation) data as a binary representation of the `JSONB` value, which eliminates whitespace, duplicate keys, and key ordering. `JSONB` supports [inverted indexes](inverted-indexes.html).
> 
> For a hands-on demonstration of storing and querying JSON data from a third-party API, see the [JSON tutorial](demo-json-support.html).
> ```

Do you want to make this a callout like you've done on inverted indexes? I actually used a plain paragraph cause I wasn't sure. Your call. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2759#-:-L8SYjgr33yevPvgK7rX:b-xdxcpp)*

**Comment by lhirata** [comment]
Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/partitioning.md, line 2 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8SVX-kf0UQxHWpI8et:-L8Uw0db89UiIAT_kn-h:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L2)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I think we need a callout that this is an enterprise feature and thus requies an enterprise licesne. Something like:

> Partitioning is an enterprise feature. To request and enable a trial or full enterprise license, see [Enterprise Licensing](enterprise-licensing.html). 
</blockquote></details>

Done.

---

*[_includes/computed-columns/partitioning.md, line 31 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8Dy43tP4uQYBCz53uM:-L8UvXPdEfrDKXtGtkrM:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/_includes/computed-columns/partitioning.md#L31)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Parens are still on a separate line, @lhirata.
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 15 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8SU417J0NbMoRs2ATZ:-L8UwLv8Aa4TtLKDaYXg:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L15)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Also, maybe make this a bulleted list:

```
- **Partitioning** ...

- **`JSONB`** ...
```


</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 15 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8SSpOT3g4f0INeATee:-L8UwctxCMuY5ICzPrms:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L15)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I'd prefer "you" over "a CockroachDB operator".

Also, in the last sentence, I don't understand `a stored computed column locality`. Can you rephrase that so that it's easier to unpack?
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 24 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8DzIOztmUQxE-VAdNw:-L8UwEmt9yP9EaQBYTa3:b-bu91xm) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/computed-columns.md#L24)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

hmm, not sure what would be better either. something like "cannot be used to compute other computed columns"?
</blockquote></details>

Updated

---

*[v2.0/jsonb.md, line 188 at r1](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8SXETsG8eNRZdtNziP:-L8UxoB62ZVveMurjbpo:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/6552f66e2dbbff74c791694430348cd1d1965212/v2.0/jsonb.md#L188)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

You might want to change the table name in this included example from `students` to something else. `students` is already used on this page, and so copy/pasting the `create table` in this example gives you an error:

```
root@:26257/test> CREATE TABLE students (                                                                                                                                             id STRING PRIMARY KEY AS (profile->>'id') STORED,                                                                                                                                 profile JSONB                                                                                                                                                                 );
pq: relation "students" already exists
```
</blockquote></details>

Done.

---

*[v2.0/jsonb.md, line 9 at r5](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8SXrp7iQsyvnEx6fho:-L8Uy3AxA7jJzIPsSgws:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/830a7e8ca0ca5c1af05f5d088956e06833a3be8a/v2.0/jsonb.md#L9)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Do you want to make this a callout like you've done on inverted indexes? I actually used a plain paragraph cause I wasn't sure. Your call. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759)*

**Comment by danhhz** [comment]
lgtm. just a couple nits and a meta-question about examples in our docs

---

Review status: 0 of 8 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/simple.md, line 9 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8XutEL64AotBFXADzC:-L8XutEL64AotBFXADzD:b-bcctkx) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/_includes/computed-columns/simple.md#L9)):*
> ```Markdown
>     first_name STRING,
>     last_name STRING,
>     full_name STRING AS (CONCAT(first_name, ' ', last_name)) STORED
> ```

hmm it's late in the process to be changing examples so I won't be offended if you ignore this, but I feel the need the point out that this doesn't internationalize well. for example, many people would not consider this their full name without it containing their middle name. (in practice, i'd advise any real application to have a single unstructured `name STRING` field)

more generally, to what extent do we care that the examples in our docs represent crl's current thoughts on schema design best practices (particularly when it's not the point of the example)?

---

*[v2.0/computed-columns.md, line 17 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8XtbrV1HeThrZYpGd_:-L8XtbrW-xS5nE-Kyv5V:bfbgch4) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L17)):*
> ```Markdown
> - **Partitioning** requires that partitions are defined using columns that are a prefix of the [primary key](primary-key.html). In the case of geo-partitioning, some applications will want to collapse the number of possible values in this column, to make certain classes of queries more performant. For example, if a users table has a country and state column, then you can make a stored computed column locality with a reduced domain for use in partitioning. For more information, see the [partitioning example](#create-a-table-with-geo-partitions-and-a-computed-column) below.
> 
> - **JSONB** columns are used for storing semi-structured `JSONB` data. When the table's primary information is stored in `JSONB`, it's useful to index a particular field of the `JSONB` document. In particular, computed columns allow for the following use case: a two-column table with a `PRIMARY KEY` column and a `payload` column, whose primary key is computed as some field from the `payload` column. This alleviates the need to manually separate your JSON blobs from your primary keys. For more information, see the [`JSONB` example](#create-a-table-with-a-jsonb-column-and-a-computed-column) below.
> ```

nit: "manually separate your primary keys from your json blobs" reads a little more clearly to me

---

*[v2.0/computed-columns.md, line 23 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8Xtu2p9MOsoj_CoVsM:-L8Xtu2p9MOsoj_CoVsN:bdgwbuy) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L23)):*
> ```Markdown
> Computed columns:
> 
> - Cannot be added once a table is created. Follow the [GitHub issue](https://github.com/cockroachdb/cockroach/issues/22652) for updates on this limitation.
> ```

nit: s/once/after/ ?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-:-L8XxDYP45hlh8dnzxQ4:b-xuj3zm)*

**Comment by jseldess** [comment]
*[_includes/computed-columns/simple.md, line 9 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8XutEL64AotBFXADzC:-L8XzWeJ1w0Sd115qnbz:b-p0i12j) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/_includes/computed-columns/simple.md#L9)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

hmm it's late in the process to be changing examples so I won't be offended if you ignore this, but I feel the need the point out that this doesn't internationalize well. for example, many people would not consider this their full name without it containing their middle name. (in practice, i'd advise any real application to have a single unstructured `name STRING` field)

more generally, to what extent do we care that the examples in our docs represent crl's current thoughts on schema design best practices (particularly when it's not the point of the example)?
</blockquote></details>

Just talked to @danhhz offline. Beyond being functionally clear, our examples should definitely aim to demonstrate best practices. That easier said than done, though, so I'll make it a 2.1 project to have Docs work with Eng to create some type of reference for us.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759)*

**Comment by danhhz** [comment]
Review status: 0 of 8 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[v2.0/computed-columns.md, line 13 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8Y7eKJ4RmX-bmdkp7i:-L8Y7eKJ4RmX-bmdkp7j:bc5mjc5) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L13)):*
> ```Markdown
> ## Why use computed columns?
> 
> Computed columns are especially useful when used with partitioning or [`JSONB`](jsonb.html) columns.
> ```

Missed mentioning this the first time... Another compelling use for these is that you can create a secondary index on a computed column. An example could be a table that is frequently sorted by the sum of some fixed number of things: sports scores (q1+q2+q3+wq4+ot) or something like that?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759)*

**Comment by danhhz** [comment]
new example lgtm, thanks!

---

Review status: 0 of 9 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[_includes/computed-columns/secondary-index.md, line 65 at r7](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8YrUP9BT_MbKakeDLV:-L8YrUP9BT_MbKakeDLW:b46huod) ([raw file](https://github.com/cockroachdb/docs/blob/cbbd721799166db66764103b084caab016a08b29/_includes/computed-columns/secondary-index.md#L65)):*
> ```Markdown
> 
> 
> table: id, home team, away team, home scores (h1+h2+ot), away scores, computed column for each score, secondary index on computed columns, compare
> ```

intentional?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-:-L8YrwHM0hmERweuSdV7:b3n2kl1)*

**Comment by lhirata** [comment]
Review status: 0 of 9 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[_includes/computed-columns/secondary-index.md, line 65 at r7](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8YrUP9BT_MbKakeDLV:-L8YsO0TCL8jyookTYCZ:b-y5qhpu) ([raw file](https://github.com/cockroachdb/docs/blob/cbbd721799166db66764103b084caab016a08b29/_includes/computed-columns/secondary-index.md#L65)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

intentional?
</blockquote></details>

whoops. deleted

---

*[_includes/computed-columns/simple.md, line 9 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8XutEL64AotBFXADzC:-L8YBWML2wWQjiKFame7:b-z1haba) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/_includes/computed-columns/simple.md#L9)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Just talked to @danhhz offline. Beyond being functionally clear, our examples should definitely aim to demonstrate best practices. That easier said than done, though, so I'll make it a 2.1 project to have Docs work with Eng to create some type of reference for us.
</blockquote></details>

won't change the example n but will note this for later!

---

*[v2.0/computed-columns.md, line 13 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8Y7eKJ4RmX-bmdkp7i:-L8YoTZ78nVJKb0cpQvR:b-l77que) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L13)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Missed mentioning this the first time... Another compelling use for these is that you can create a secondary index on a computed column. An example could be a table that is frequently sorted by the sum of some fixed number of things: sports scores (q1+q2+q3+wq4+ot) or something like that?
</blockquote></details>

Added example - Dan, can you take a look?

---

*[v2.0/computed-columns.md, line 17 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8XtbrV1HeThrZYpGd_:-L8YogjU-6IDd-VfYXTf:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L17)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "manually separate your primary keys from your json blobs" reads a little more clearly to me
</blockquote></details>

Done.

---

*[v2.0/computed-columns.md, line 23 at r6](https://beta.reviewable.io/reviews/cockroachdb/docs/2759#-L8Xtu2p9MOsoj_CoVsM:-L8YokYJ8IAZtZJZk21b:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e93814eb855ced4c2861beff2e9c836bc7d20b3f/v2.0/computed-columns.md#L23)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: s/once/after/ ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2759)*
